File: logic/src/main/java/com/tp/holdem/logic/model/PokerTable.java
package com.tp.holdem.logic.model;

import com.tp.holdem.logic.HandOperations;
import com.tp.holdem.logic.PlayerExceptions;
import com.tp.holdem.logic.PlayerFunctions;
import com.tp.holdem.model.common.Moves;
import com.tp.holdem.model.common.Phase;
import com.tp.holdem.model.message.dto.PokerTableDTO;
import io.vavr.collection.HashMap;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import io.vavr.control.Option;
import lombok.Builder;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;

import java.util.Objects;
import java.util.concurrent.atomic.AtomicLong;

@Slf4j
@Value
@Builder(toBuilder = true)
public class PokerTable {
	private int smallBlindAmount;
	private int bigBlindAmount;
	private boolean showdown;
	private boolean gameOver;
	@Builder.Default
	private Deck deck = Deck.brandNew();
	@Builder.Default
	private Phase phase = Phase.START;
	@Builder.Default
	private PlayerNumber bettingPlayer = PlayerNumber.empty();
	@Builder.Default
	private PlayerNumber winnerPlayer = PlayerNumber.empty();
	@Builder.Default
	private PlayerNumber dealer = PlayerNumber.empty();
	@Builder.Default
	private PlayerNumber bigBlind = PlayerNumber.empty();
	@Builder.Default
	private PlayerNumber smallBlind = PlayerNumber.empty();
	@Builder.Default
	private List<Player> allPlayers = List.empty();
	@Builder.Default
	private List<Card> cardsOnTable = List.empty();
	@Builder.Default
	private Map<Integer, Moves> movesThisPhase = HashMap.empty();

	public PokerTable addPlayer(Player player) {
		return this.toBuilder()
				.allPlayers(getAllPlayers().append(player))
				.build();
	}

	public PokerTable playerLeft(PlayerNumber playerNumber) {
		final Player foundPlayer = allPlayers
				.find(PlayerFunctions.byNumber(playerNumber.getNumber()))
				.getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND);

		final List<Player> modifiedPlayers = allPlayers
				.replace(foundPlayer, foundPlayer.toBuilder().inGame(false).build());

		return this.toBuilder().allPlayers(modifiedPlayers).build();
	}

	public int highestBetThisPhase() {
		return allPlayers.map(Player::getBetAmountThisPhase).max().getOrElse(0);
	}

	public int potAmount() {
		return allPlayers.map(Player::getBetAmount).sum().intValue();
	}

	public int potAmountThisPhase() {
		return allPlayers.map(Player::getBetAmountThisPhase).sum().intValue();
	}

	public PokerTable nextPhase() {
		final Phase nextPhase = phase.nextPhase();
		if (nextPhase == Phase.PRE_FLOP)
			return goToPreFlopPhase();
		if (nextPhase == Phase.FLOP)
			return goToNextPhase(3);
		if (nextPhase == Phase.TURN || nextPhase == Phase.RIVER)
			return goToNextPhase(1);
		throw new IllegalStateException("There is no next phase");
	}

	private PokerTable goToPreFlopPhase() {
		final Player smallBlindPlayer = getSmallBlind().getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND);
		final Player bettingPlayer = PlayerFunctions.FIRST_BET_IN_ROUND.apply(smallBlindPlayer, this);

		return this.toBuilder()
				.phase(getPhase().nextPhase())
				.allPlayers(getAllPlayers().replace(smallBlindPlayer, bettingPlayer))
				.bettingPlayer(PlayerNumber.of(bettingPlayer.getNumber()))
				.movesThisPhase(HashMap.empty())
				.build();
	}

	private PokerTable goToNextPhase(int cards) {
		final List<Player> preparedPlayers = getAllPlayers().map(Player::prepareForNewPhase);
		final Player smallBlindPlayer = preparedPlayers
				.find(player -> Objects.equals(player.getNumber(), getSmallBlind().getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND).getNumber()))
				.getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND);

		final PokerTable updatedTable = this.toBuilder()
				.phase(getPhase().nextPhase())
				.allPlayers(preparedPlayers)
				.cardsOnTable(getCardsOnTable().appendAll(deck.drawCards(cards)))
				.movesThisPhase(HashMap.empty())
				.build();

		final Player bettingPlayer = PlayerFunctions.BET_IN_PHASE.apply(smallBlindPlayer, updatedTable);

		return updatedTable.toBuilder()
				.allPlayers(updatedTable.getAllPlayers().replace(smallBlindPlayer, bettingPlayer))
				.bettingPlayer(PlayerNumber.of(bettingPlayer.getNumber()))
				.build();
	}

	public PhaseStatus phaseStatus() {
		int notAllInCount = allPlayers
				.filter(Player::playing)
				.count(player -> !player.isAllIn());

		boolean allPlayersMoved = movesThisPhase.size() == allPlayers.filter(Player::playing).size();

		if (notAllInCount <= 1 && allPlayersMoved)
			return PhaseStatus.EVERYBODY_ALL_IN;

		int notFoldedCount = allPlayers
				.filter(Player::playing)
				.count(player -> !player.isFolded());

		if (notFoldedCount == 1)
			return PhaseStatus.EVERYBODY_FOLDED;

		boolean allBetsAreEqual = allPlayers
				.filter(Player::playing)
				.forAll(player -> player.isAllIn() || player.getBetAmountThisPhase() == highestBetThisPhase());

		return allPlayersMoved && allBetsAreEqual ? PhaseStatus.READY_FOR_NEXT : PhaseStatus.KEEP_GOING;
	}

	public PokerTable nextPlayerToBet() {
		final Player newBettingPlayer = allPlayers.get((allPlayers.indexOf(getBettingPlayer().getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND)) + 1) % allPlayers.size());
		final Player modifiedNewBettingPlayer = PlayerFunctions.BET_IN_PHASE.apply(newBettingPlayer, this);

		return this.toBuilder()
				.allPlayers(allPlayers.replace(newBettingPlayer, modifiedNewBettingPlayer))
				.bettingPlayer(PlayerNumber.of(modifiedNewBettingPlayer.getNumber()))
				.build();
	}

	public PokerTable roundOver() {
		final List<Player> playersAfterRound = allPlayers.map(Player::roundOver);

		final PokerTable updatedTable = this.toBuilder()
				.allPlayers(playersAfterRound)
				.build();

		final Player winner = HandOperations.findWinner(playersAfterRound, updatedTable);

		final Player prizedWinner = winner.toBuilder()
				.chipsAmount(winner.getChipsAmount() + updatedTable.potAmount())
				.build();

		return updatedTable.toBuilder()
				.winnerPlayer(PlayerNumber.of(winner.getNumber()))
				.allPlayers(updatedTable.getAllPlayers().replace(winner, prizedWinner))
				.phase(Phase.OVER)
				.build();
	}

	public PokerTable dealCards() {
		log.debug("Dealing cards to players");
		final List<Player> playersWithCards = deck.dealCards(2, allPlayers);
		return this.toBuilder()
				.allPlayers(playersWithCards)
				.build();
	}

	public PokerTable preparePlayersForNewGame(int startingChips) {
		return toBuilder()
				.allPlayers(getAllPlayers().map(player -> player.prepareForNewGame(startingChips)))
				.build();
	}

	public Option<Player> getBettingPlayer() {
		return allPlayers.find(PlayerFunctions.byNumber(bettingPlayer.getNumber()));
	}

	public Option<Player> getWinnerPlayer() {
		return allPlayers.find(PlayerFunctions.byNumber(winnerPlayer.getNumber()));
	}

	public Option<Player> getDealer() {
		return allPlayers.find(PlayerFunctions.byNumber(dealer.getNumber()));
	}

	public Option<Player> getBigBlind() {
		return allPlayers.find(PlayerFunctions.byNumber(bigBlind.getNumber()));
	}

	public Option<Player> getSmallBlind() {
		return allPlayers.find(PlayerFunctions.byNumber(smallBlind.getNumber()));
	}

	public PokerTableDTO toDTO() {
		return PokerTableDTO.builder()
				.phase(phase)
				.gameOver(gameOver)
				.potAmount(potAmount())
				.potAmountThisPhase(potAmountThisPhase())
				.smallBlindAmount(smallBlindAmount)
				.bigBlindAmount(bigBlindAmount)
				.dealer(getDealer().map(Player::toPlayerDTO).getOrNull())
				.bigBlind(getBigBlind().map(Player::toPlayerDTO).getOrNull())
				.smallBlind(getSmallBlind().map(Player::toPlayerDTO).getOrNull())
				.allPlayers(allPlayers.map(Player::toPlayerDTO).toJavaList())
				.bettingPlayer(getBettingPlayer().map(Player::toPlayerDTO).getOrNull())
				.cardsOnTable(cardsOnTable.map(Card::toDTO).toJavaList())
				.winnerPlayer(getWinnerPlayer().map(Player::toPlayerDTO).getOrNull())
				.build();
	}

	public PokerTable newRound(AtomicLong handCount) {
		log.debug("Preparing players for new round");
		final List<Player> playersWithCleanBets = getAllPlayers().map(Player::prepareForNewRound);

		final Player dealerPlayer = playersWithCleanBets.get((int) (handCount.get() % playersWithCleanBets.size()));

		final Player smallBlindPlayer = playersWithCleanBets.get((int) ((handCount.get() + 1) % playersWithCleanBets.size()));
		log.debug(String.format("Taking small blind from player: %d", smallBlindPlayer.getNumber()));
		final Player newSmallBlindPlayer = PlayerFunctions.SMALL_BLIND_TIME.apply(smallBlindPlayer, this);

		final Player bigBlindPlayer = playersWithCleanBets.get((int) ((handCount.get() + 2) % playersWithCleanBets.size()));
		log.debug(String.format("Taking big blind from player: %d", bigBlindPlayer.getNumber()));
		final Player newBigBlindPlayer = PlayerFunctions.BIG_BLIND_TIME.apply(bigBlindPlayer, this);

		final PokerTable updatedTable = toBuilder()
				.deck(Deck.brandNew())
				.cardsOnTable(List.empty())
				.phase(Phase.START)
				.showdown(false)
				.movesThisPhase(HashMap.empty())
				.allPlayers(playersWithCleanBets.replace(smallBlindPlayer, newSmallBlindPlayer).replace(bigBlindPlayer, newBigBlindPlayer))
				.winnerPlayer(PlayerNumber.empty())
				.bettingPlayer(PlayerNumber.empty())
				.dealer(PlayerNumber.of(dealerPlayer.getNumber()))
				.bigBlind(PlayerNumber.of(newBigBlindPlayer.getNumber()))
				.smallBlind(PlayerNumber.of(newSmallBlindPlayer.getNumber()))
				.build();

		return updatedTable.dealCards();
	}

	public PokerTable playerMove(Integer playerNumber, Moves move, int betAmount) {
		final Player actionPlayer = getAllPlayers()
				.find(PlayerFunctions.byNumber(playerNumber))
				.getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND);

		final Player playerAfterAction;

		switch (move) {
			case FOLD:
				playerAfterAction = actionPlayer.fold();
				break;
			case ALLIN:
				playerAfterAction = actionPlayer.allIn();
				break;
			case BET:
				playerAfterAction = actionPlayer.bet(betAmount);
				break;
			case CHECK:
				playerAfterAction = actionPlayer;
				break;
			case CALL:
				playerAfterAction = actionPlayer.bet(highestBetThisPhase() - actionPlayer.getBetAmountThisPhase());
				break;
			case RAISE:
				playerAfterAction = actionPlayer.bet(highestBetThisPhase() - actionPlayer.getBetAmountThisPhase()).bet(betAmount);
				break;
			default:
				throw new IllegalArgumentException(String.format("Unsupported action type: %s", move));
		}

		return this.toBuilder()
				.allPlayers(allPlayers.replace(actionPlayer, playerAfterAction.bettingTurnOver()))
				.movesThisPhase(movesThisPhase.put(actionPlayer.getNumber(), move))
				.build();
	}

	public boolean isNotPlayable() {
		return allPlayers
				.filter(Player::isInGame)
				.filter(player -> player.getChipsAmount() > 0)
				.length() < 2;
	}

	public PokerTable gameOver() {
		return toBuilder()
				.gameOver(true)
				.phase(Phase.OVER)
				.allPlayers(getAllPlayers().map(Player::gameOver))
				.build();
	}
}

File: logic/src/main/java/com/tp/holdem/logic/model/PokerTable.kt
package com.tp.holdem.logic.model

import com.tp.holdem.logic.HandOperations
import com.tp.holdem.logic.PlayerBettingFunctions
import com.tp.holdem.logic.PlayerExceptions
import com.tp.holdem.logic.PlayerFinders
import com.tp.holdem.model.common.Moves
import com.tp.holdem.model.common.Phase
import com.tp.holdem.model.message.dto.PokerTableDTO
import io.vavr.collection.HashMap
import io.vavr.collection.List
import io.vavr.collection.Map
import io.vavr.control.Option
import java.util.concurrent.atomic.AtomicLong

data class PokerTable(
        val smallBlindAmount: Int = 0,
        val bigBlindAmount: Int = 0,
        val showdown: Boolean = false,
        val gameOver: Boolean = false,
        val deck: Deck = Deck.brandNew(),
        val phase: Phase = Phase.START,
        private val bettingPlayer: PlayerNumber = PlayerNumber.empty(),
        private val winnerPlayer: PlayerNumber = PlayerNumber.empty(),
        private val dealer: PlayerNumber = PlayerNumber.empty(),
        private val bigBlind: PlayerNumber = PlayerNumber.empty(),
        private val smallBlind: PlayerNumber = PlayerNumber.empty(),
        val allPlayers: List<Player> = List.empty<Player>(),
        val cardsOnTable: List<Card> = List.empty<Card>(),
        val movesThisPhase: Map<Int, Moves> = HashMap.empty()
) {
    companion object {
        val log = org.slf4j.LoggerFactory.getLogger(PokerTable.javaClass)

        @JvmStatic
        fun withBlinds(bigBlindAmount: Int, smallBlindAmount: Int): PokerTable {
            return PokerTable(bigBlindAmount = bigBlindAmount, smallBlindAmount = smallBlindAmount)
        }
    }

    val isNotPlayable: Boolean
        get() = allPlayers
                .filter { it.inGame }
                .filter { player -> player.chipsAmount > 0 }
                .length() < 2

    fun addPlayer(player: Player): PokerTable {
        return this.copy(
                allPlayers = allPlayers.append(player)
        )
    }

    fun playerLeft(playerNumber: PlayerNumber): PokerTable {
        val foundPlayer = allPlayers
                .find(PlayerFinders.byNumber(playerNumber.number))
                .getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND)

        val modifiedPlayers = allPlayers
                .replace(foundPlayer, foundPlayer.copy(inGame = false))

        return this.copy(
                allPlayers = modifiedPlayers
        )
    }

    fun highestBetThisPhase(): Int {
        return allPlayers.map { it.betAmountThisPhase }.max().getOrElse(0)
    }

    fun potAmount(): Int {
        return allPlayers.map { it.betAmount }.sum().toInt()
    }

    fun potAmountThisPhase(): Int {
        return allPlayers.map { it.betAmountThisPhase }.sum().toInt()
    }

    fun nextPhase(): PokerTable {
        val nextPhase = phase.nextPhase()
        if (nextPhase == Phase.PRE_FLOP)
            return goToPreFlopPhase()
        if (nextPhase == Phase.FLOP)
            return goToNextPhase(3)
        if (nextPhase == Phase.TURN || nextPhase == Phase.RIVER)
            return goToNextPhase(1)
        throw IllegalStateException("There is no next phase")
    }

    private fun goToPreFlopPhase(): PokerTable {
        val dealerPlayer = getDealer().getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND)
        val bettingPlayer = PlayerBettingFunctions.firstBetInRound(dealerPlayer, this)

        return this.copy(
                phase = phase.nextPhase(),
                allPlayers = allPlayers.replace(dealerPlayer, bettingPlayer),
                bettingPlayer = PlayerNumber.of(bettingPlayer.number),
                movesThisPhase = HashMap.empty<Int, Moves>()
        )
    }

    private fun goToNextPhase(cards: Int): PokerTable {
        val preparedPlayers = allPlayers.map<Player> { it.prepareForNewPhase() }

        val updatedTable = this.copy(
                phase = phase.nextPhase(),
                allPlayers = preparedPlayers,
                cardsOnTable = cardsOnTable.appendAll(deck.drawCards(cards)),
                movesThisPhase = movesThisPhase.filterValues { it.goingToNextPhase() }
        )

        return updatedTable.nextPlayerToBetAfter(bigBlind.number)
    }

    fun phaseStatus(): PhaseStatus {
        val notAllInCount = allPlayers
                .filter { it.inGame }
                .count { player -> !player.allIn }

        val allPlayersMoved = movesThisPhase.size() == allPlayers.filter { it.inGame }.size()

        if (notAllInCount <= 1 && allPlayersMoved)
            return PhaseStatus.EVERYBODY_ALL_IN

        val notFoldedCount = allPlayers
                .filter { it.inGame }
                .count { player -> !player.folded }

        if (notFoldedCount == 1)
            return PhaseStatus.EVERYBODY_FOLDED

        val allBetsAreEqual = allPlayers
                .filter { it.playing() }
                .forAll { player -> player.allIn || player.betAmountThisPhase == highestBetThisPhase() }

        return if (allPlayersMoved && allBetsAreEqual) PhaseStatus.READY_FOR_NEXT else PhaseStatus.KEEP_GOING
    }

    fun nextPlayerToBet(): PokerTable {
        return nextPlayerToBetAfter(bettingPlayer.number)
    }

    private fun nextPlayerToBetAfter(playerNumber: Int?): PokerTable {
        log.debug(String.format("Finding next to bet after: %d", playerNumber))

        val bettingPlayerIndex = allPlayers.indexWhere(PlayerFinders.byNumber(playerNumber))
        var newBettingPlayer: Player
        var count = 1
        do {
            newBettingPlayer = allPlayers.get((bettingPlayerIndex + count++) % allPlayers.size())
        } while (newBettingPlayer.folded)

        val modifiedNewBettingPlayer = PlayerBettingFunctions.betInPhase(newBettingPlayer, this)

        return this.copy(
                allPlayers = allPlayers.replace(newBettingPlayer, modifiedNewBettingPlayer),
                bettingPlayer = PlayerNumber.of(modifiedNewBettingPlayer.number)
        )
    }

    fun roundOver(): PokerTable {
        val playersAfterRound = allPlayers.map<Player> { it.roundOver() }

        val updatedTable = this.copy(
                allPlayers = playersAfterRound
        )

        val winner = HandOperations.findWinner(playersAfterRound, updatedTable)

        val prizedWinner = winner.copy(
                chipsAmount = winner.chipsAmount + updatedTable.potAmount()
        )

        return this.copy(
                winnerPlayer = PlayerNumber.of(winner.number),
                allPlayers = updatedTable.allPlayers.replace(winner, prizedWinner),
                phase = Phase.OVER
        )
    }

    fun dealCards(): PokerTable {
        log.debug("Dealing cards to players")
        val playersWithCards = deck.dealCards(2, allPlayers)
        return this.copy(
                allPlayers = playersWithCards
        )
    }

    fun preparePlayersForNewGame(startingChips: Int): PokerTable {
        return this.copy(
                allPlayers = allPlayers.map { player -> player.prepareForNewGame(startingChips) }
        )
    }

    fun getBettingPlayer(): Option<Player> {
        return allPlayers.find(PlayerFinders.byNumber(bettingPlayer.number))
    }

    fun getWinnerPlayer(): Option<Player> {
        return allPlayers.find(PlayerFinders.byNumber(winnerPlayer.number))
    }

    fun getDealer(): Option<Player> {
        return allPlayers.find(PlayerFinders.byNumber(dealer.number))
    }

    fun getBigBlind(): Option<Player> {
        return allPlayers.find(PlayerFinders.byNumber(bigBlind.number))
    }

    fun getSmallBlind(): Option<Player> {
        return allPlayers.find(PlayerFinders.byNumber(smallBlind.number))
    }

    fun toDTO(): PokerTableDTO {
        return PokerTableDTO.builder()
                .phase(phase)
                .gameOver(gameOver)
                .potAmount(potAmount())
                .potAmountThisPhase(potAmountThisPhase())
                .smallBlindAmount(smallBlindAmount)
                .bigBlindAmount(bigBlindAmount)
                .dealer(getDealer().map { it.toPlayerDTO() }.orNull)
                .bigBlind(getBigBlind().map { it.toPlayerDTO() }.orNull)
                .smallBlind(getSmallBlind().map { it.toPlayerDTO() }.orNull)
                .allPlayers(allPlayers.map { it.toPlayerDTO() }.toJavaList())
                .bettingPlayer(getBettingPlayer().map { it.toPlayerDTO() }.orNull)
                .cardsOnTable(cardsOnTable.map { it.toDTO() }.toJavaList())
                .winnerPlayer(getWinnerPlayer().map { it.toPlayerDTO() }.orNull)
                .build()
    }

    fun newRound(handCount: AtomicLong): PokerTable {
        log.debug("Preparing players for new round")
        val playersWithCleanBets = allPlayers.map { it.prepareForNewRound() }

        val smallBlindPlayer = playersWithCleanBets.get(((handCount.get() + 1) % playersWithCleanBets.size()).toInt())
        log.debug(String.format("Taking small blind from player: %d", smallBlindPlayer.number))
        val newSmallBlindPlayer = PlayerBettingFunctions.smallBlindTime(smallBlindPlayer, this)

        val dealerPlayer: Player
        if (allPlayers.size() == 2)
            dealerPlayer = newSmallBlindPlayer
        else
            dealerPlayer = playersWithCleanBets.get((handCount.get() % playersWithCleanBets.size()).toInt())

        val bigBlindPlayer: Player
        if (allPlayers.size() == 2)
            bigBlindPlayer = playersWithCleanBets.get((handCount.get() % playersWithCleanBets.size()).toInt())
        else
            bigBlindPlayer = playersWithCleanBets.get(((handCount.get() + 2) % playersWithCleanBets.size()).toInt())

        log.debug(String.format("Taking big blind from player: %d", bigBlindPlayer.number))
        val newBigBlindPlayer = PlayerBettingFunctions.bigBlindTime(bigBlindPlayer, this)

        val updatedTable = this.copy(
                deck = Deck.brandNew(),
                cardsOnTable = List.empty(),
                phase = Phase.START,
                showdown = false,
                movesThisPhase = HashMap.empty<Int, Moves>(),
                allPlayers = playersWithCleanBets.replace(smallBlindPlayer, newSmallBlindPlayer).replace(bigBlindPlayer, newBigBlindPlayer),
                winnerPlayer = PlayerNumber.empty(),
                bettingPlayer = PlayerNumber.empty(),
                dealer = PlayerNumber.of(dealerPlayer.number),
                bigBlind = PlayerNumber.of(newBigBlindPlayer.number),
                smallBlind = PlayerNumber.of(newSmallBlindPlayer.number)
        )

        return updatedTable.dealCards()
    }

    fun playerMove(playerNumber: Int?, move: Moves, betAmount: Int): PokerTable {
        val actionPlayer = allPlayers
                .find(PlayerFinders.byNumber(playerNumber))
                .getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND)

        val playerAfterAction: Player

        when (move) {
            Moves.FOLD -> playerAfterAction = actionPlayer.fold()
            Moves.ALLIN -> playerAfterAction = actionPlayer.allIn()
            Moves.BET -> playerAfterAction = actionPlayer.bet(betAmount)
            Moves.CHECK -> playerAfterAction = actionPlayer
            Moves.CALL -> playerAfterAction = actionPlayer.bet(highestBetThisPhase() - actionPlayer.betAmountThisPhase)
            Moves.RAISE -> playerAfterAction = actionPlayer.bet(highestBetThisPhase() - actionPlayer.betAmountThisPhase).bet(betAmount)
            else -> throw IllegalArgumentException(String.format("Unsupported action type: %s", move))
        }

        return this.copy(
                allPlayers = allPlayers.replace(actionPlayer, playerAfterAction.bettingTurnOver()),
                movesThisPhase = movesThisPhase.put(actionPlayer.number, move)
        )
    }

    fun gameOver(): PokerTable {
        return this.copy(
                gameOver = true,
                phase = Phase.OVER,
                allPlayers = allPlayers.map { it.gameOver() }
        )
    }

    fun showdownMode(): PokerTable {
        return this.copy(
                showdown = true
        )
    }
}